p2p/service.py

```diff
@@ -32,28 +35,47 @@ def __init__(self, token: CancelToken=None) -> None:
         Once _run() returns, set the finished event, call cleanup() and
         finished_callback (if one was passed).
         """
+        if self.is_running:
+            raise RuntimeError("Cannot start the service while it's already running")
+        elif self.is_finished:
+            raise RuntimeError("Cannot restart a service after it has completed")
+
         try:
-            await self._run()
+            async with self._run_lock:
+                await self._run()
         except OperationCancelled as e:
             self.logger.info("%s finished: %s", self, e)
         except Exception:
             self.logger.exception("Unexpected error in %s, exiting", self)
         else:
             self.logger.debug("%s finished cleanly", self)
         finally:
-            # Set self.finished before anything else so that other coroutines started by this
-            # service exit while we wait for cleanup().
-            self.finished.set()
-            await self.cleanup()
+            try:
+                await self.cleanup()
+            except Exception:
+                self.logger.exception("Unexepected error during cleanup in %s", self)
+                raise
+
             if finished_callback is not None:
                 finished_callback(self)
 
     async def cleanup(self) -> None:
         """Run the service's _cleanup() coroutine."""
+        # Set self.finished before anything else so that other coroutines started by this
+        # service exit while we wait for cleanup().
+        self.finished.set()
```

Comment:
I think it'd make more sense to leave this at the end of `run()` as it is not really part of any cleanup